### PR TITLE
Add PyG remote backend example.

### PR DESCRIPTION
**Pull request by coledie**
- [x] Forked repo is synced with upstream -> github shows no code delta outside of the desired.
    https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork
- [x] Tests are passing? https://github.com/microsoft/DeepGNN/blob/main/CONTRIBUTING.md#run-tests
- [x] Changelog and documentation updated.
- [x] PR is labeled using the label menu on the right side.

New Behavior
----------------
Add example with PyG remote backend example.

### CODE REVIEW

File: examples/pytorch/pyg/train.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    model = GraphSAGE(
+        in_channels=50,
+        hidden_channels=64,
+        num_layers=2,
+        out_channels=64,
+    ).to(device)
+    optimizer = torch.optim.Adam(model.parameters(), lr=0.005)
+
+    for epoch in range(1, 6):
+        loss = train()
```
Comment (mortid0) [inline examples/pytorch/pyg/train.py:157]: Please, add a a test() method similar [to this](https://github.com/pyg-team/pytorch_geometric/blob/decec47e3c9cd67b8ddbd58a15edd00de0c530ef/examples/graph_sage_unsup.py#L57C1-L67C29)(with evaluation only on a test dataset) and an assert to make sure we get expected values from training.

File: examples/pytorch/pyg/test_interface.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import argparse
+import torch
+
+from examples.pytorch.pyg.train import DeepGNNFeatureStore, DeepGNNGraphStore
+from deepgnn.graph_engine import SamplingStrategy
+from deepgnn.graph_engine.data.citation import Cora
+from torch_geometric.loader import LinkNeighborLoader
+
+
+def test_interface():
```
Comment (mortid0) [inline examples/pytorch/pyg/test_interface.py:19]: This is not needed, just use an assert in the train.py

File: examples/pytorch/pyg/BUILD [hunk 3]
```diff
@@ -0,0 +1,52 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+load("@rules_python//python:defs.bzl", "py_test", "py_library")
+load("@pip_deps//:requirements.bzl", "requirement")
+
+py_library(
```
Comment (mortid0) [inline examples/pytorch/pyg/BUILD:7]: 1. it should be a binary, so you can run it as an example.
2. Rename train.py to main.py. This is a Bazel convention I think.

File: examples/pytorch/pyg/test_interface.py (hunk 2 above)
Comment (coledie) [reply examples/pytorch/pyg/test_interface.py:19]: Train.py uses ppi which is too slow to run in CI pipeline.

Comment (mortid0) [reply examples/pytorch/pyg/test_interface.py:19]: what stops you from using a smaller graph, which will be also faster to load?

Comment (coledie) [reply examples/pytorch/pyg/test_interface.py:19]: We want it to be a useful example that shows actual learning

Comment (mortid0) [reply examples/pytorch/pyg/test_interface.py:19]: > We want it to be a useful example that shows actual learning. Cora is not big enough.

Why many papers use it for benchmarks?

> Also none of our other examples do testing like this

[Of course we do](https://github.com/microsoft/DeepGNN/blob/1e35f511e6f18857e1aa0c774e9dee23b9adc88c/.github/workflows/ci.yml#L154C9-L155C73). There is no assert in main, we only verify code runs without exceptions and have UT to verify logic. It will be easier to add an assert in this train.py and skip UTs, because you are not adding any new code outside of examples folder.

File: examples/pytorch/pyg/train.py (hunk 1 above)
Comment (coledie) [reply examples/pytorch/pyg/train.py]: Is it possible with Cora since all edges have type 0? It would just be the training set accuracy which maybe is okay also?

File: examples/pytorch/pyg/main.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        """To be implemented by :class:`GraphStore` subclasses."""
+        edge_type = int(edge_attr.edge_type[1])
+        edge = self.ge.sample_edges(
+            self.ge.edge_count(0), np.array(edge_type), SamplingStrategy.Random
+        )
+        edge_index = (
+            torch.Tensor(edge[:, 0]).long(),
+            torch.Tensor(edge[:, 1]).long(),
+        )
+        # assert False, edge_index
```
Comment (mortid0) [inline examples/pytorch/pyg/main.py:89]: Please, remove comments.

File: examples/pytorch/pyg/train.py (hunk 1 above)
Comment (coledie) [reply examples/pytorch/pyg/train.py]: I added the test but even though loss is going down, test stays random, do you see anything wrong with it?

File: examples/pytorch/pyg/BUILD [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        "//src/python/deepgnn/pytorch/encoding/twinbert:deepgnn_pytorch_encoding_twinbert",
+        "//src/python/deepgnn/pytorch/modeling:deepgnn_pytorch_modeling",
+        "//src/python/deepgnn/pytorch/encoding/twinbert/deepspeed:deepgnn_pytorch_encoding_twinbert_ds",
+        "//src/python/deepgnn/pytorch/nn:deepgnn_pytorch_nn",
+        "//src/python/deepgnn/pytorch/training:deepgnn_pytorch_training",
+        "//examples/pytorch:example_torch_tests",
+    ],
+)
+
+py_test(
```
Comment (mortid0) [inline examples/pytorch/pyg/BUILD:29]: This is not needed anymore

File: examples/pytorch/pyg/main.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
+"""PyGeo interface training example.
+
+https://github.com/pyg-team/pytorch_geometric/blob/ae84a38f14591ba9b8ce64e704e04ea1271c3b78/examples/graph_sage_unsup_ppi.py#L12
```
Comment (mortid0) [inline examples/pytorch/pyg/main.py:3]: This comment doesn't seem to be relevant anymore, because you are using Cora.

File: examples/pytorch/pyg/main.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+class DeepGNNFeatureStore(FeatureStore):
+    """An abstract base class to access features from a remote feature store."""
+
+    def __init__(self, ge):
+        """Initialize DeepGNN feature store."""
+        super().__init__()
+        self.ge = ge
+
+    def _put_tensor(self, tensor, attr) -> bool:
+        """To be implemented by :class:`FeatureStore` subclasses."""
```
Comment (mortid0) [inline examples/pytorch/pyg/main.py:30]: This docstring is not needed. Same below.

File: examples/pytorch/pyg/main.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+        for i in ["x", "y"]:
+            ta = TensorAttr()
+            ta.group_name = "0"
+            ta.attr_name = f"{i}"
+            output.append(ta)
+        return output
+
+
+class DeepGNNGraphStore(GraphStore):
+    """An abstract base class to access edges from a remote graph store.
```
Comment (mortid0) [inline examples/pytorch/pyg/main.py:59]: Please update docstrings, DeepGNNGraphStore is definitely not abstract.

File: examples/pytorch/pyg/main.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+        ),
+        batch_size=140,
+        shuffle=True,
+        neg_sampling_ratio=1.0,
+        num_neighbors=[5, 5],
+        num_workers=1,
+        persistent_workers=True,
+        edge_label_index=(
+            ("0", "0", "0"),
+            torch.Tensor(ge.sample_edges(2708, np.array(0), SamplingStrategy.Random))
```
Comment (mortid0) [inline examples/pytorch/pyg/main.py:153]: Please, extract a const for this magic number, 2708.

File: examples/pytorch/pyg/main.py [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+        num_layers=2,
+        out_channels=7,
+    ).to(device)
+    optimizer = torch.optim.Adam(model.parameters(), lr=0.005)
+
+    for epoch in range(1, 10):
+        loss = train()
+        test_f1 = test()
+        print(f"Epoch: {epoch:03d}, Loss: {loss:.4f}, Test F1: {test_f1:.4f}")
+    assert loss <= 0.55
```
Comment (mortid0) [inline examples/pytorch/pyg/main.py:188]: You want to assert f1 value on test dataset, not loss.

File: examples/pytorch/pyg/main.py [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+        hidden_channels=64,
+        num_layers=2,
+        out_channels=7,
+    ).to(device)
+    optimizer = torch.optim.Adam(model.parameters(), lr=0.005)
+
+    for epoch in range(1, 10):
+        loss = train()
+        test_f1 = test()
+        print(f"Epoch: {epoch:03d}, Loss: {loss:.4f}, Test F1: {test_f1:.4f}")
```
Comment (mortid0) [inline examples/pytorch/pyg/main.py:187]: Please, use logger instead of prints

File: examples/pytorch/pyg/main.py [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+        optimizer.step()
+
+        total_loss += float(loss) * link_pred.numel()
+        total_examples += link_pred.numel()
+
+    return total_loss / total_examples
+
+
+def test():
+    for data in tqdm.tqdm(loader):
```
Comment (mortid0) [inline examples/pytorch/pyg/main.py:130]: forgot `model.eval()`

File: examples/pytorch/pyg/main.py [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+        """Obtain all edge attributes stored in the :class:`GraphStore`."""
+        output = []
+        ta = EdgeAttr(("0", "0", "0"), "coo", size=[self.ge.node_count(0), 2708])
+        output.append(ta)
+
+        return output
+
+
+def train():
+    """Train the model."""
```
Comment (mortid0) [inline examples/pytorch/pyg/main.py:104]: Self explanatory docstring. Either remove here or add to the test function below.

File: examples/pytorch/pyg/main.py (hunk 6 above)
Comment (coledie) [reply examples/pytorch/pyg/main.py:3]: I am using their graphsage model still and most of their training loop.

File: examples/pytorch/pyg/main.py (hunk 10 above)
Comment (coledie) [reply examples/pytorch/pyg/main.py:188]: See other comment above, I need help with this.

File: examples/pytorch/pyg/BUILD (hunk 5 above)
Comment (coledie) [reply examples/pytorch/pyg/BUILD:29]: The whole build file?

File: examples/pytorch/pyg/main.py (hunk 6 above)
Comment (mortid0) [reply examples/pytorch/pyg/main.py:3]: Training loop is the same across many models. Please, remove this comment.

File: examples/pytorch/pyg/train.py (hunk 1 above)
Comment (mortid0) [reply examples/pytorch/pyg/train.py]: Could you explain, what is random? f1 score stays low or non deterministic? If the former, you need to debug outputs of wrappers and compare them to relevant PyG tensors. If the latter, you need to set random seed. Either [globally](https://github.com/microsoft/DeepGNN/blob/1e35f511e6f18857e1aa0c774e9dee23b9adc88c/src/python/deepgnn/pytorch/common/utils.py#L31), or in the calls to sampling methods.

File: examples/pytorch/pyg/main.py [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+            ta.attr_name = f"{i}"
+            output.append(ta)
+        return output
+
+
+class DeepGNNGraphStore(GraphStore):
+    """A class to access edges from a remote graph store.
+
+    Args:
+        edge_attr_cls (EdgeAttr, optional): A user-defined
```
Comment (mortid0) [inline examples/pytorch/pyg/main.py:58]: This is a wrong docstring

Comment (mortid0) [reply examples/pytorch/pyg/main.py:58]: There is no `edge_attr_cls` argument

File: examples/pytorch/pyg/main.py (hunk 9 above)
Comment (mortid0) [reply examples/pytorch/pyg/main.py:159]: 2708 is still not extracted as const. I believe this should be number of edges of type 0 or something like that.

File: .github/workflows/ci.yml [hunk 15]
```diff
@@ -145,6 +145,8 @@ jobs:
           python-version: ${{ matrix.python-version }}
       - name: Install torch
         run: pip install --timeout 300 torch==1.13.1
+      - name: Pip freeze
```
Comment (mortid0) [inline .github/workflows/ci.yml:148]: why is this needed?

**Review by mortid0 — APPROVED** [review]

File: .github/workflows/ci.yml (hunk 15 above)
Comment (coledie) [reply .github/workflows/ci.yml:148]: Debug, it looks like the CI is failing because of new PyG version, this will hopefully get resolved with Ray fix so I will finalize then.

**Review by mortid0 — APPROVED** [review]
